sdk/typescript/src/transactions/ObjectCache.ts

```diff
@@ -0,0 +1,316 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import { bcs } from '../bcs/index.js';
+import type { SuiClient } from '../client/client.js';
+import type { ExecuteTransactionBlockParams } from '../client/index.js';
+import type { Signer } from '../cryptography/keypair.js';
+import { normalizeSuiAddress } from '../utils/sui-types.js';
+import type { OpenMoveTypeSignature } from './blockData/v2.js';
+import type { TransactionBlock } from './TransactionBlock.js';
+import type { TransactionBlockDataResolverPlugin } from './TransactionBlockDataResolver.js';
+
+export interface ObjectCacheEntry {
+	objectId: string;
+	version: string;
+	digest: string;
+	owner: string | null;
+	initialSharedVersion: string | null;
+}
+
+export interface MoveFunctionCacheEntry {
+	package: string;
+	module: string;
+	function: string;
+	parameters: OpenMoveTypeSignature[];
+}
+
+export interface CacheEntryTypes {
+	OwnedObject: ObjectCacheEntry;
+	SharedObject: ObjectCacheEntry;
+	MoveFunction: MoveFunctionCacheEntry;
+}
+export abstract class AsyncCache {
+	protected abstract get<T extends keyof CacheEntryTypes>(
+		type: T,
+		key: string,
+	): Promise<CacheEntryTypes[T] | null>;
+	protected abstract set<T extends keyof CacheEntryTypes>(
+		type: T,
+		key: string,
+		value: CacheEntryTypes[T],
+	): Promise<void>;
+	protected abstract delete<T extends keyof CacheEntryTypes>(type: T, key: string): Promise<void>;
+	abstract clear<T extends keyof CacheEntryTypes>(type?: T): Promise<void>;
+
+	async getObject(id: string) {
+		const [owned, shared] = await Promise.all([
+			this.get('OwnedObject', id),
+			this.get('SharedObject', id),
+		]);
+
+		return owned ?? shared ?? null;
+	}
+
+	async getObjects(ids: string[]) {
+		return Promise.all([
+			...ids.map((id) => this.get('OwnedObject', id)),
+			...ids.map((id) => this.get('SharedObject', id)),
```

Comment:
This works fine currently, but will have extra nulls so index in results don't map indexes in the id list, I should probably fix that